File: 200-number-of-islands/number-of-islands.go
func numIslands(grid [][]byte) int {
    row := len(grid)
    col := len(grid[0])

    num := 0

    for i:=0 ; i < row ; i++ {
        for j:=0 ; j< col ; j++{
            if grid[i][j] == '1' {
                dfs(grid,i,j,row, col)
                num++
            }
        }
    }
    return num
}

func dfs(grid [][]byte, i int, j int, row int, col int) {
    if i>=row || i<0 || j>=col || j<0 {
        return
    }

    // mark current visited
    grid[i][j] = '2'

    // visit top
    if i>0 && grid[i-1][j] == '1' {
        dfs(grid,i-1,j,row,col)
    }

    // visit bottom
    if i<row-1 && grid[i+1][j] == '1' {
        dfs(grid,i+1,j,row,col)
    }

    // visit left
    if j>0 && grid[i][j-1] == '1' {
        dfs(grid,i,j-1,row,col)
    }

    // visit right
    if j<col-1 && grid[i][j+1] == '1' {
        dfs(grid,i,j+1,row,col)
    }
}

File: 200-number-of-islands/number-of-islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        # for each grid
        #   if i'ts not visited yet
        #   visit everything it's connected to
        #   island +=1
        row = len(grid)
        col = len(grid[0])

        islands = 0

        for r in range(row):
            for c in range(col):
                if grid[r][c] == "1":
                    self.visitIsland(grid,r,c)
                    islands+=1
        return islands


    def visitIsland(self, grid: List[List[str]], r : int, c : int) :
        row = len(grid)
        col = len(grid[0])
        
        if r<0 or r>=row or c<0 or c>=col or grid[r][c] == "0":
            return
        grid[r][c] = "0"
        self.visitIsland(grid, r-1,c)
        self.visitIsland(grid, r+1,c)
        self.visitIsland(grid, r,c+1)
        self.visitIsland(grid, r,c-1)